src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -1778,19 +1800,116 @@ RelationRestrictionPartitionKeyIndex(RelationRestriction *relationRestriction)
 		partitionKeyTargetAttrIndex++;
 
 		if (!targetEntry->resjunk &&
-			IsA(targetExpression, Var) &&
-			IsPartitionColumn(targetExpression, relationPlannerParseQuery))
+			IsA(targetExpression, Var))
 		{
 			Var *targetColumn = (Var *) targetExpression;
 
-			if (targetColumn->varno == relationRestriction->index)
+			if (!IsPartitionColumn(targetExpression, relationPlannerParseQuery))
+			{
+				continue;
+			}
+
+			RangeTblEntry *resRTE = FindRelationRTEForColumn(
+				relationPlannerParseQuery->rtable, targetColumn);
+			if (resRTE->rtekind == RTE_RELATION && GetRTEIdentity(resRTE) ==
+				findRteIdentityContext->targetRTEIdentity)
+			{
+#if PG_VERSION_NUM < PG_VERSION_13
+
+/*
+ * Before pg13, targetColumn will contain the varno and varattno pointing to the
+ * JOIN_RTE, so we find the column that points to the RELATION_RTE.
+ */
+				Oid relationId = InvalidOid;
+				Var *column = NULL;
+				FindReferencedTableColumn(targetExpression, NIL,
+										  relationPlannerParseQuery, &relationId,
+										  &column);
+				targetColumn = column;
+#endif
+
+				*partitionKeyIndex = partitionKeyTargetAttrIndex;
+				return (Var *) copyObject(targetColumn);
+			}
+		}
+	}
+
+	return NULL;
+}
+
+
+/*
+ * FindRelationRTEForColumn finds the relation RTE for the given targetColumn from
+ * the given rtable list.
+ */
+static RangeTblEntry *
+FindRelationRTEForColumn(List *rtable, Var *targetColumn)
+{
+	RangeTblEntry *resRTE = (RangeTblEntry *) list_nth(rtable, targetColumn->varno - 1);
+#if PG_VERSION_NUM <= PG_VERSION_13
+
+/*
+ * Before pg13, join columns would point to RTE_JOIN rtes, but we
+ * want to find RELATION_RTE, so we look at the joinaliasvars to
+ * find the RELATION_RTE for the targetColumn.
+ */
+	if (resRTE->rtekind == RTE_JOIN)
+	{
+		Var *joinaliasvar = NULL;
+		int joinvarattno = 1;
+		foreach_ptr(joinaliasvar, resRTE->joinaliasvars)
+		{
+			if (joinvarattno == targetColumn->varattno)
 			{
-				return partitionKeyTargetAttrIndex;
+				resRTE = (RangeTblEntry *) list_nth(rtable, joinaliasvar->varno - 1);
+				break;
 			}
+			joinvarattno++;
 		}
 	}
+#endif
+	return resRTE;
+}
+
+
+/*
+ * FindQueryContainingRTEIdentity walks on the given node to find a query
+ * which has an RTE that has a given rteIdentity.
+ */
+static bool
+FindQueryContainingRTEIdentity(Node *node, FindRteIdentityContext *context)
```

Comment:
A high level API might be nicer: `Query * q= FindQueryContainingRTEIdentity(Query *mainQuery, rteIndex)` where we create the context and call `FindQueryContainingRTEIdentity`

Reply:
I named the high level function as `FindQueryContainingRteIdentity` and the internal method with a `Internal` suffix. 

Mainly because I think `Query` is more correct than `SubQuery` and when we have `FindQueryContainingRteIdentity` and `FindSubQueryContainingRteIdentity` it seems a bit confusing to see which one to use